Fix linker flags for building Python package for conda.

chrisdembia opened the pull request:
Fixes issue #2771 

### Brief summary of changes

This PR brings over a fix I saw elsewhere on GitHub for getting the python package to work with Anaconda's python on Mac.

The diff in this PR provides more background.

### Testing I've completed

On Mac, I built the Python package with Anaconda and ran the python CTest tests.

### CHANGELOG.md (choose one)

- updated...

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/opensim-org/opensim-core/2773)
<!-- Reviewable:end -->

aymanhab commented:
Nice investigative work @chrisdembia 👍 The question in my mind is whether it's worth it to complicate the build system and future maintenance to support this specific use-case, when it sounds like many other environment including miniconda work out of the box. Do you know if this is Mac specific, and if Anaconda works out of the box on windows? Thank you

chrisdembia commented:
> Do you know if this is Mac specific, ...

I think this issue could occur on any UNIXy system where the Python interpreter statically links to the Python library, because the PYTHON_LIBRARIES CMake variable assumes that the Python interpreter *dynamically* links to the Python library. From my experience and from what I've read on GitHub, this occurs with Macs and Anaconda Python, but the fix I implemented does not assume Mac or Anaconda; it attempts to get closer to the root problem.

The fix I implemented is copied from the links I put in the CMakeLists.txt file. It is likely that there are better fixes. The python interpreter itself provides some C/C++ flags that should be used when building C/C++ extensions. I briefly looked to see if we could leverage those flags to replace PYTHON_LIBRARIES and avoid adding a special case, but nothing stood out to me. I could investigate more deeply, but other projects have already chosen this fix.

I think using Anaconda Python on Mac is a substantial use case that we should support because Conda (which uses Anaconda Python) will be the easiest way for us to package the OpenSim Python package. In fact, I ran into this same bug myself when trying to package Moco for Conda over the last few days.

> ...and if Anaconda works out of the box on windows?
I am also attempting to package Moco for Conda for Windows but am running into unrelated issues, so I do not know yet. I do not know yet if Anaconda's Python interpreter links statically or dynamically to the Python library, or if doing so would present a similar issue on Windows. My guess is not.

> (including miniconda?)

I do not know enough about the differences between Miniconda and Anaconda, but my guess is the issue is the same with Miniconda.

> whether it's worth it to complicate the build system and future maintenance

I've tried to give enough background links so that if the issue evolves in the future, it'll be easy to see how other projects are adapting.

aymanhab commented:
Thanks @chrisdembia for the clarification of the scope. Merging.
